### ISSUE TITLE
Require metrics to specify which categories must be enabled when they register

### ISSUE THREAD

**Issue by natduca** [catapult-project/catapult#1818]
We want to be able to get traces reliably that can do railscoring. To do this, I think we probably want all of our tracepoints required for railscoring behind a single trace category.

The hard part here is that railscoring may be too narrow a category name. We have ongoing discussions about a more general approach here --- for instance, user percieved performance ["upp"], or system_health. If we think of railscoring as a leading edge of that effort, then maybe the right label is upp, or system_health or system_health_metrics.

**Comment by natduca** [comment]
@primiano how do the metrics required for memory infra system health plan get triggered in traceconfig?

**Comment by benshayden** [comment]
This is blocked on the backlog entry "Come up with branding story for system_metrics, upp, rail etc"
We'll send a survey to pick a name in Q1.

**Comment by primiano** [comment]
Somehow missed this thread. Via "memory_dump_config" in the traceconfig JSON (well actually today we have to support also the old style category filter as telemetry still uses that one)

See https://code.google.com/p/chromium/codesearch#chromium/src/base/trace_event/trace_config.h&l=117

**Comment by benshayden** [comment]
When tracing categories are saved in the trace metadata, then the metrics side panel etc. can avoid offering a metric if its required categories were not enabled.

**Comment by benshayden** [comment]
This is low priority but it should be easy and will have a moderate impact on the usability of the metrics side panel and metric health. I'm not sure when I'll be able to get to it.

**Comment by natduca** [comment]
If this happens, we should probably do some design discussions with nedn etc

**Comment by petrcermak** [comment]
+@nedn: We could use this in benchmarks (so that they wouldn't have to specify which tracing categories to enable)

**Comment by petrcermak** [comment]
One thing to keep in mind: Some metrics can work with events from various categories. For example, the memory metric requires <tt>disabled-by-default-memory-infra</tt>, but it can also use data from <tt>disabled-by-default-memory-infra.v8.code_stats</tt> when available.

**Comment by benshayden** [comment]
If we're going to be thinking about the categories that were enabled for a given trace, then we need them to always be logged in the trace metadata: https://bugs.chromium.org/p/chromium/issues/detail?id=638756

**Comment by nedn** [comment]
I think let's punt on this until PCU and how the trace events should be tagged is clearer?

**Comment by benshayden** [comment]
Should this be archived as unmotivated?

**Comment by eakuefner** [comment]
I think so. Please reopen if we want to consider this at this time.

**Comment by benshayden** [comment]
https://codereview.chromium.org/2717713003/

**Comment by benshayden** [comment]
> > https://codereview.chromium.org/2717713003/diff/1/tools/perf/benchmarks/metric_dependencies.py#newcode11
> > > tools/perf/benchmarks/metric_dependencies.py:11: _CATEGORIES_FOR_METRICS = {
> > > ulan: instead of this, I think we should make a all_metrics.json file in
> > catapult/tracing/tracing/metrics that list out all the metrics & their dependent
> > tracing categories. 
> > > 
> > > The reason for this is it would allow the metric side panel in about://tracing
> > to also use this dependency map, so we can do things like: not displaying
> > certain metrics in the panel if the trace does not have the required categories,
> > show people that they will need to enable which categories to get the metric.
> > 
> > Instead of creating a new kind of json file, can we register the categories as
> > metadata when the metrics are registered in the MetricRegistry like how
> > supportsRangeOfInterest is registered?
> > MetricRegistry can require metrics to specify requiredCategories in their
> > metadata, as part of the existing all_metrics presubmit, instead of creating a
> > new presubmit and new infrastructure.
> 
> The reason is prefer the JSON file is:
> 1) the dependencies will be needed by both JS (tracing) & Telemetry (Python). If that is configured by pure JS, we need another way to invoke d8 to extract the categories, which seems less ideal.
> 2) We already talk about the need for a centralized all_metrics.json file to track all the metrics, historgrams & making sure that they are uniquely identified. So if such a file is needed anyway, why not also use it for tracking the categories? We can make s.t like:
> { 
>    'metric_foo': {
>       'histograms': {'foo_1':.., 'foo_2',...':..,}
>       'required_trace_categories': [...],
>     },
>    'metric_bar': {...},
> }

@nedn @eakuefner 

Is it ok if we continue this discussion here?

Thanks for clarifying!

But, sorry, can you say more about why telemetry couldn't use a few lines to call vinn to collate the metric metadata from JS?

It seems to me like a nice feature for metrics to be able to describe everything about themselves in a single file so they don't forget to update other files when they update the metric. It's already easy enough to forget to add files to trace_viewer.gypi and all_metrics.html, and the presubmits to check them are complex enough that I'd like to see if we can avoid complicating them further by requiring manual updates to requiredCategories and histogramNames in another file.

I don't know how much you've used histograms.xml, but it's not uncommon for people to forget to add/update/remove entries when they add/update/remove UMA macros. At least, I did, and I saw many other people do so, too. I've seen bugs asking people to update it and CLs scrubbing it, all of which seems like a pain that is avoidable with our metrics. Maybe they solved it with a presubmit since I last used it, but it seems like it might be nice to avoid that complexity if we can.

Or maybe you see an easy way to ensure that requiredCategories in all_metrics.json is updated when metrics are modified to depend on different events?
Or maybe you see some reason that it isn't as big of a problem as it seems to me?

**Comment by nedn** [comment]
@tdresser & @ulan  for this discussion.

I think the your argument about the `requiredCategories` should be close to the metric code that depend on trace events from those categories is a fair point. We can have Telemetry call vinn to collate the metric metadata from JS. 

I still think we should have an all_metrics.json file as a high level interface for understanding all the metrics & histograms they produce. My problem with metric_X.html is that the file is already very dense with the implementation logic & is not fit for certain kind of metadata about the metrics & histograms. e.g: looking at https://github.com/catapult-project/catapult/blob/master/tracing/tracing/metrics/v8/runtime_stats_metric.html, it's unclear to me which histograms the metrics would produce because the creation logic is done dynamically.

Having said that, in this case, specifying `requiredCategories` in the metric registry code sgtm.

**Comment by ulan** [comment]
Both approaches seem fine to me. Summarizing the discussion.

Categories in the json:
+ declarative
+ centralized
+ fast to parse in python
- can get out of sync with metric.

Categories in the js:
+ expressive (full power of JS)
+ closer to metric code
- slower to get in python (requires running js in d8/vinn)
- can get out of sync with metric

Note that I put "can get out of sync" in both approaches because even in JS we cannot currently verify that the set of categories listed in `requiredCategories` is complete. I had this problem when working on the EQT metric: I used the existing function tr.e.chrome.getInteractiveTimestamps which implicitly relies on "loading" category. Note that getInteractiveTimestamps is defined in another file than the EQT metric.  If someone changes the function to depend on another category, the `requiredCategories` in the  EQT metric would get out of sync.

We can somewhat mitigate that problem with a dynamic check when running the metric. If the model had a list of categories that were used for recording the trace, then getInteractiveTimestamps could at least assert model.checkTraceConfigIncludes("loading"). This would work for both json and js file approaches.

**Comment by tdresser** [comment]
Could we declare the dependencies in the js, and then have a presubmit check, which generates all_metrics.json?

It's a bit annoying to have generated files checked in, but it would get us some of the benefits (being closer to the metric code) of being written in JS, along with some of the benefits (fast to parse in python) of the json approach.

**Comment by natduca** [comment]
Fwiw, we've been here before 5 years ago with an metrics.json and it burned us. @anniesullie may remember. We wanted to move to this model of the code is where thigns are documented precisely because of endless drift. Maybe things are better now, idk. I like code-is-the-truth model but I'm a driveby opinion holder not a stakeholder. Talk to Annie for more.

**Comment by anniesullie** [comment]
We've talked through this problem a lot over the past few months! I've continued to feel strongly about keeping things documented in the code after the many mistakes we've made in the past with a centralized model (unit-info.json, dashboard sheriffing, etc). Would be happy to pick up the discussion again.

@simonhatch FYI

**Comment by nedn** [comment]
I think "centralization of documentation is bad" is too general of an assessment. For example, devtool team also uses https://chromium.googlesource.com/chromium/src/+/master/third_party/WebKit/Source/core/inspector/browser_protocol.json  as a declarative way to keep track of their public APsI and I find it make it easy for me as the user to look for APIs that I care about. 

My main concern with relying only on code is that `tracing/tracing/metric/` has not done a great jobs of declaring which histograms each metric will produce. On one hand, that requires metric users to read the implementation to understand the API (bad) & on another hand, we risks having people produce zillion of histograms dynamically by mistake.

**Comment by tdresser** [comment]
Doesn't declaring the dependencies in the js and then having a presubmit check which generates all_metrics.json solve most of the issues being raised?

**Comment by nedn** [comment]
I am sorry for conflating two different issues in one bug. The two issues are:
1) We need a way to declare which tracing categories a metric depend on.
2) We may need a all_metrics.json file that declare which histograms each metric produces & further documentation.

For (1), I am totally supportive with declaring the dependencies in metrics code. Which tracing categories a metric depend on is an implementation details that I think people should need not to care about in general. There is also no need to generate all_metrics.json since calling to vinn to execute JS does not have terrible overhead AFAIK.

(2) should be continuously discussed in a different bug. I remember Tim filed a bug about ensuring metrics' histograms should be unique but can't find it.

**Comment by tdresser** [comment]
Has there been any progress here?

The topic of histogram metadata comes up pretty frequently. I'm still hoping to eventually have a mapping between UMA histograms and catapult histograms. This is getting somewhat more complicated as we get more complex breakdown trees, and more structured data on the server side via UKM.

**Comment by benshayden** [comment]
This bug appears to be about allowing/requiring TBMv2 metrics to specify which tracing categories they require. I started that here: https://chromium-review.googlesource.com/c/catapult/+/695743

I'm not sure I see how that relates to mapping between UMA metrics and telemetry metrics?

**Comment by tdresser** [comment]
Sorry, the thread that connects them is the idea of metric metadata. The required categories are metadata, and related UMA metrics are also metadata.

You're right that this is off topic though. We're following up via email.

**Comment by benshayden** [comment]
This bug is old and basically fixed now, so I'm going to close it. Let's open new bugs if there are still issues.
